batch/src/main/java/org/openmrs/analytics/JdbcFetchUtil.java

```diff
@@ -0,0 +1,205 @@
+			    JdbcIO.<KV<String, Integer>, String> readAll().withDataSourceConfiguration(this.dataSourceConfig)
+			            .withParameterSetter(new JdbcIO.PreparedStatementSetter<KV<String, Integer>>() {
+				            
+				            @Override
+				            public void setParameters(KV<String, Integer> element, PreparedStatement preparedStatement)
+				                    throws Exception {
+					            String[] range = Objects.requireNonNull(element.getKey()).split(",");
+					            preparedStatement.setInt(1, Integer.parseInt(range[0]));
+					            preparedStatement.setInt(2, Integer.parseInt(range[1]));
+				            }
+			            }).withOutputParallelization(false)
+			            .withQuery(
+			                String.format("select uuid from %s where %s >= ? and %s < ?", this.tableName, tableId, tableId))
+			            .withRowMapper(new JdbcIO.RowMapper<String>() {
+				            
+				            @Override
+				            public String mapRow(ResultSet resultSet) throws Exception {
+					            return resultSet.getString("uuid");
+				            }
+			            }).withCoder(NullableCoder.of(StringUtf8Coder.of())));
+		}
+	}
+	
+	/**
+	 * batch together request using batchSize and generate segment descriptors `_id?<uuid>,<uuid>..`
+	 */
+	public static class CreateSearchSegments extends PTransform<PCollection<String>, PCollection<SearchSegmentDescriptor>> {
+		
+		private final String baseBundleUrl;
+		
+		private final String resourceType;
+		
+		private final int batchSize;
+		
+		public CreateSearchSegments(String resourceType, String baseBundleUrl, int batchSize) {
+			this.baseBundleUrl = baseBundleUrl;
+			this.batchSize = batchSize;
+			this.resourceType = resourceType;
+		}
+		
+		@Override
+		public PCollection<SearchSegmentDescriptor> expand(PCollection<String> uuIds) {
+			return uuIds
+			        // create KV required by GroupIntoBatches
+			        .apply(ParDo.of(new DoFn<String, KV<String, String>>() {
+				        
+				        @ProcessElement
+				        public void processElement(ProcessContext c) {
+					        if (c.element() != null) {
+						        c.output(KV.of(resourceType, c.element()));
+					        }
+				        }
+			        }))
+			        .apply(String.format("GroupIntoBatches of %s", this.batchSize),
+			            GroupIntoBatches.<String, String> ofSize(this.batchSize))
+			        .apply("Generate Segments", ParDo.of(new DoFn<KV<String, Iterable<String>>, SearchSegmentDescriptor>() {
+				        
+				        @ProcessElement
+				        public void processElement(@Element KV<String, Iterable<String>> element,
+				                OutputReceiver<SearchSegmentDescriptor> r) {
+					        List<String> uuIds = new ArrayList<String>();
+					        element.getValue().forEach(uuIds::add);
+					        r.output(SearchSegmentDescriptor.create(
+					            String.format("%s?_id=%s", baseBundleUrl, String.join(",", uuIds)), uuIds.size()));
```

Comment:
I don't think that setting the `count` of `SearchSegmentDescriptor` matters in this case? If yes, we can either change `FhirSearchUtil` and avoid setting `count()` when the count is negative (and use `-1` here) OR just add a comment here clarifying that setting `count` here has no effect.

Reply:
True! I have made a comment, thanks!